File: tests/hex4/parse_hex4.py
#!/usr/bin/env python3
"""Independent HEX4 binary parser used by the regression harness."""

from __future__ import annotations

import json
import struct
import sys
from pathlib import Path
from typing import Any

from hex4_common import (
    HEADER_SIZE,
    LAYER_SCALES,
    RECORD_SIZE,
    canonical_json,
    layer_center_axial,
)


HEADER_STRUCT = struct.Struct("<4siifffii")
COUNT_STRUCT = struct.Struct("<I")
RECORD_STRUCT = struct.Struct("<bbHhhhBBBBBx")


class Hex4ParseError(ValueError):
    pass


def _need(data: bytes, offset: int, size: int, label: str) -> None:
    if offset + size > len(data):
        raise Hex4ParseError(f"truncated while reading {label}: need {offset + size} bytes, have {len(data)}")


def parse_bytes(data: bytes, *, strict: bool = True) -> dict[str, Any]:
    _need(data, 0, HEADER_SIZE, "header")
    sig, sx, sy, min_z, max_z, scale_f, cq, cr = HEADER_STRUCT.unpack_from(data, 0)
    if sig != b"HEX4":
        raise Hex4ParseError(f"invalid signature {sig!r}")
    if scale_f == 0.0:
        raise Hex4ParseError("invalid zero height scale")

    offset = HEADER_SIZE
    layers: list[list[dict[str, Any]]] = []
    raw_counts: list[int] = []

    for layer_index, layer_scale in enumerate(LAYER_SCALES):
        _need(data, offset, COUNT_STRUCT.size, f"layer {layer_index} count")
        (count,) = COUNT_STRUCT.unpack_from(data, offset)
        raw_counts.append(count)
        offset += COUNT_STRUCT.size

        lcq, lcr = layer_center_axial(sx, sy, layer_scale)
        layer: list[dict[str, Any]] = []
        for record_index in range(count):
            _need(data, offset, RECORD_SIZE, f"layer {layer_index} record {record_index}")
            dq, dr, h_scaled, d1, d2, d3, s1, s2, s3, pnx, pnz = RECORD_STRUCT.unpack_from(data, offset)
            h = min_z + (h_scaled / scale_f)
            layer.append(
                {
                    "dq": dq,
                    "dr": dr,
                    "q": lcq + dq,
                    "r": lcr + dr,
                    "hScaled": h_scaled,
                    "h": h,
                    "deltas": [d1, d2, d3],
                    "slopes": [s1, s2, s3],
                    "norm": [pnx, pnz],
                }
            )
            offset += RECORD_SIZE
        layers.append(layer)

    if strict and offset != len(data):
        raise Hex4ParseError(f"trailing bytes after HEX4 payload: consumed {offset}, file has {len(data)}")

    return {
        "sx": sx,
        "sy": sy,
        "header": {
            "signature": "HEX4",
            "minZ": min_z,
            "maxZ": max_z,
            "scale": scale_f,
            "cq": cq,
            "cr": cr,
        },
        "stats": {"min": min_z, "max": max_z, "avg": (min_z + max_z) / 2.0, "base": min_z},
        "center": {"q": 0, "r": 0},
        "layerScales": list(LAYER_SCALES),
        "layerCounts": raw_counts,
        "layers": layers,
        "consumedBytes": offset,
        "byteLength": len(data),
    }


def parse_file(path: str | Path, *, strict: bool = True) -> dict[str, Any]:
    return parse_bytes(Path(path).read_bytes(), strict=strict)


def main(argv: list[str] | None = None) -> int:
    argv = argv if argv is not None else sys.argv[1:]
    if not argv:
        print("usage: parse_hex4.py PATH", file=sys.stderr)
        return 2
    try:
        parsed = parse_file(argv[0], strict="--allow-trailing" not in argv[1:])
    except Exception as exc:
        print(f"parse_hex4.py: {exc}", file=sys.stderr)
        return 1
    print(canonical_json(parsed))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/hex4/parse_hex4.mjs
#!/usr/bin/env node
// Line-faithful port of frontend/app/tile_worker.js parseBinaryV3().

import fs from 'node:fs';

const SECTOR_WIDTH_METERS = 819.2;
const UNIT_HEX_PX = 32.0;
const METERS_PER_PIXEL = 0.2;
const UNIT_HEX_WIDTH_METERS = UNIT_HEX_PX * METERS_PER_PIXEL;

function worldToAxialScale(x, y, s) {
    const h = UNIT_HEX_WIDTH_METERS * s;
    const A = (Math.sqrt(3) / 2) * h;
    const q = x / A;
    const r = (y - (q * 0.5 * h)) / h;
    return { q, r };
}

function parseBinaryV3(buffer) {
    const view = new DataView(buffer);
    const sig = String.fromCharCode(view.getUint8(0), view.getUint8(1), view.getUint8(2), view.getUint8(3));
    if (sig !== 'HEX4') throw new Error("Invalid Sig");

    const sx = view.getInt32(4, true);
    const sy = view.getInt32(8, true);
    const minZ = view.getFloat32(12, true);
    const maxZ = view.getFloat32(16, true);
    const scale = view.getFloat32(20, true);
    const cq = view.getInt32(24, true);
    const cr = view.getInt32(28, true);

    let offset = 32;
    const layers = [];
    const layerCounts = [];
    const scales = [24.0, 6.0, 3.0, 1.0];

    const minX = sx * SECTOR_WIDTH_METERS;
    const minY = sy * SECTOR_WIDTH_METERS;
    const cenX = minX + SECTOR_WIDTH_METERS * 0.5;
    const cenY = minY + SECTOR_WIDTH_METERS * 0.5;

    for (let l = 0; l < 4; l++) {
        const count = view.getUint32(offset, true);
        layerCounts.push(count);
        offset += 4;
        const layer = [];
        const sc = scales[l];
        const rawC = worldToAxialScale(cenX, cenY, sc);
        const lcq = Math.round(rawC.q);
        const lcr = Math.round(rawC.r);

        for (let i = 0; i < count; i++) {
            const dq = view.getInt8(offset);
            const dr = view.getInt8(offset + 1);
            const hn = view.getUint16(offset + 2, true);
            const d1 = view.getInt16(offset + 4, true);
            const d2 = view.getInt16(offset + 6, true);
            const d3 = view.getInt16(offset + 8, true);
            const s1 = view.getUint8(offset + 10);
            const s2 = view.getUint8(offset + 11);
            const s3 = view.getUint8(offset + 12);
            const nx = view.getUint8(offset + 13);
            const nz = view.getUint8(offset + 14);
            offset += 16;

            layer.push({
                dq, dr,
                q: lcq + dq, r: lcr + dr,
                hScaled: hn,
                h: minZ + (hn / scale),
                deltas: [d1, d2, d3],
                slopes: [s1, s2, s3],
                norm: [nx, nz]
            });
        }
        layers.push(layer);
    }

    return {
        sx, sy,
        header: { signature: sig, minZ, maxZ, scale, cq, cr },
        stats: { min: minZ, max: maxZ, avg: (minZ + maxZ) / 2, base: minZ },
        center: { q: 0, r: 0 },
        layerScales: scales,
        layerCounts,
        layers,
        consumedBytes: offset,
        byteLength: buffer.byteLength
    };
}

function canonicalJson(value) {
    if (value === null || typeof value !== 'object') {
        return JSON.stringify(value);
    }
    if (Array.isArray(value)) {
        return '[' + value.map(canonicalJson).join(',') + ']';
    }
    const keys = Object.keys(value).sort();
    return '{' + keys.map((key) => JSON.stringify(key) + ':' + canonicalJson(value[key])).join(',') + '}';
}

function main() {
    const path = process.argv[2];
    if (!path) {
        console.error('usage: parse_hex4.mjs PATH');
        return 2;
    }
    try {
        const buf = fs.readFileSync(path);
        const arrayBuffer = buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
        console.log(canonicalJson(parseBinaryV3(arrayBuffer)));
        return 0;
    } catch (err) {
        console.error(`parse_hex4.mjs: ${err.message}`);
        return 1;
    }
}

process.exitCode = main();
